File: kotlin/src/test/java/it/gabrieletondi/telldontaskkata/useCase/OrderApprovalUseCaseTest.kt
package it.gabrieletondi.telldontaskkata.useCase

import it.gabrieletondi.telldontaskkata.domain.Order
import it.gabrieletondi.telldontaskkata.domain.OrderStatus
import it.gabrieletondi.telldontaskkata.doubles.TestOrderRepository
import org.hamcrest.Matchers.`is`
import org.hamcrest.Matchers.nullValue
import org.junit.Assert.assertThat
import org.junit.Test

class OrderApprovalUseCaseTest {
    private val orderRepository: TestOrderRepository = TestOrderRepository()
    private val useCase = OrderApprovalUseCase(orderRepository)

    @Test
    @Throws(Exception::class)
    fun approvedExistingOrder() {
        val initialOrder = Order()
        initialOrder.status = OrderStatus.CREATED
        initialOrder.id = 1
        orderRepository.addOrder(initialOrder)
        val request = OrderApprovalRequest()
        request.orderId = 1
        request.isApproved = true
        useCase.run(request)
        val savedOrder: Order = orderRepository.savedOrder!!
        assertThat(savedOrder.status, `is`(OrderStatus.APPROVED))
    }

    @Test
    @Throws(Exception::class)
    fun rejectedExistingOrder() {
        val initialOrder = Order()
        initialOrder.status = OrderStatus.CREATED
        initialOrder.id = 1
        orderRepository.addOrder(initialOrder)
        val request = OrderApprovalRequest()
        request.orderId = 1
        request.isApproved = false
        useCase.run(request)
        val savedOrder: Order = orderRepository.savedOrder!!
        assertThat(savedOrder.status, `is`(OrderStatus.REJECTED))
    }

    @Test(expected = RejectedOrderCannotBeApprovedException::class)
    @Throws(Exception::class)
    fun cannotApproveRejectedOrder() {
        val initialOrder = Order()
        initialOrder.status = OrderStatus.REJECTED
        initialOrder.id = 1
        orderRepository.addOrder(initialOrder)
        val request = OrderApprovalRequest()
        request.orderId = 1
        request.isApproved = true
        useCase.run(request)
        assertThat(orderRepository.savedOrder, `is`(nullValue()))
    }

    @Test(expected = ApprovedOrderCannotBeRejectedException::class)
    @Throws(Exception::class)
    fun cannotRejectApprovedOrder() {
        val initialOrder = Order()
        initialOrder.status = OrderStatus.APPROVED
        initialOrder.id = 1
        orderRepository.addOrder(initialOrder)
        val request = OrderApprovalRequest()
        request.orderId = 1
        request.isApproved = false
        useCase.run(request)
        assertThat(orderRepository.savedOrder, `is`(nullValue()))
    }

    @Test(expected = ShippedOrdersCannotBeChangedException::class)
    @Throws(Exception::class)
    fun shippedOrdersCannotBeApproved() {
        val initialOrder = Order()
        initialOrder.status = OrderStatus.SHIPPED
        initialOrder.id = 1
        orderRepository.addOrder(initialOrder)
        val request = OrderApprovalRequest()
        request.orderId = 1
        request.isApproved = true
        useCase.run(request)
        assertThat(orderRepository.savedOrder, `is`(nullValue()))
    }

    @Test(expected = ShippedOrdersCannotBeChangedException::class)
    @Throws(Exception::class)
    fun shippedOrdersCannotBeRejected() {
        val initialOrder = Order()
        initialOrder.status = OrderStatus.SHIPPED
        initialOrder.id = 1
        orderRepository.addOrder(initialOrder)
        val request = OrderApprovalRequest()
        request.orderId = 1
        request.isApproved = false
        useCase.run(request)
        assertThat(orderRepository.savedOrder, `is`(nullValue()))
    }
}

File: csharpcore/TellDontAskKata/TellDontAskKata.Tests/UseCase/OrderApprovalUseCaseTest.cs
using System;
using TellDontAskKata.Main.Domain;
using TellDontAskKata.Main.UseCase;
using TellDontAskKata.Tests.Doubles;
using Xunit;

namespace TellDontAskKata.Tests.UseCase
{
    public class OrderApprovalUseCaseTest
    {
        private readonly TestOrderRepository _orderRepository;
        private readonly OrderApprovalUseCase _useCase;

        public OrderApprovalUseCaseTest()
        {
            _orderRepository = new TestOrderRepository();
            _useCase = new OrderApprovalUseCase(_orderRepository);
        }


        [Fact]
        public void ApprovedExistingOrder()
        {
            var initialOrder = new Order
            {
                Status = OrderStatus.Created,
                Id = 1
            };
            _orderRepository.AddOrder(initialOrder);

            var request = new OrderApprovalRequest
            {
                OrderId = 1,
                Approved = true
            };

            _useCase.Run(request);

            var savedOrder = _orderRepository.GetSavedOrder();
            Assert.Equal(OrderStatus.Approved, savedOrder.Status);
        }

        [Fact]
        public void RejectedExistingOrder()
        {
            var initialOrder = new Order
            {
                Status = OrderStatus.Created,
                Id = 1
            };
            _orderRepository.AddOrder(initialOrder);

            var request = new OrderApprovalRequest
            {
                OrderId = 1,
                Approved = false
            };

            _useCase.Run(request);

            var savedOrder = _orderRepository.GetSavedOrder();
            Assert.Equal(OrderStatus.Rejected, savedOrder.Status);
        }


        [Fact]
        public void CannotApproveRejectedOrder()
        {
            var initialOrder = new Order
            {
                Status = OrderStatus.Rejected,
                Id = 1
            };
            _orderRepository.AddOrder(initialOrder);

            var request = new OrderApprovalRequest
            {
                OrderId = 1,
                Approved = true
            };


            Action actionToTest = () => _useCase.Run(request);
      
            Assert.Throws<RejectedOrderCannotBeApprovedException>(actionToTest);
            Assert.Null(_orderRepository.GetSavedOrder());
        }

        [Fact]
        public void CannotRejectApprovedOrder()
        {
            var initialOrder = new Order
            {
                Status = OrderStatus.Approved,
                Id = 1
            };
            _orderRepository.AddOrder(initialOrder);

            var request = new OrderApprovalRequest
            {
                OrderId = 1,
                Approved = false
            };


            Action actionToTest = () => _useCase.Run(request);
            
            Assert.Throws<ApprovedOrderCannotBeRejectedException>(actionToTest);
            Assert.Null(_orderRepository.GetSavedOrder());
        }

        [Fact]
        public void ShippedOrdersCannotBeRejected()
        {
            var initialOrder = new Order
            {
                Status = OrderStatus.Shipped,
                Id = 1
            };
            _orderRepository.AddOrder(initialOrder);

            var request = new OrderApprovalRequest
            {
                OrderId = 1,
                Approved = false
            };


            Action actionToTest = () => _useCase.Run(request);

            Assert.Throws<ShippedOrdersCannotBeChangedException>(actionToTest);
            Assert.Null(_orderRepository.GetSavedOrder());
        }

    }
}
